### PR TITLE
feature: implemented ngx.req module and its function add_header().

### PR DESCRIPTION
**Pull request by spacewander**
The `ngx.req.add_header` works mostly like `ngx.req.set_header`.
However, unlike `ngx.req.set_header`, this method appends new header to the old
one instead of overriding it if the header is not builtin.

For the builtin header, this method overrides it instead of appending to it.
IMHO, it is reasonable since appending to the builtin header like `User-Agent`
is confusing and semantically wrong.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-core project.

### CODE REVIEW

**Comment by spacewander** [comment]
CC @agentzh @thibaultcha

File: lib/resty/core/request.lua [hunk 1]
```diff
@@ -310,16 +314,53 @@ do
 
         local rc
         if not value then
-            rc = C.ngx_http_lua_ffi_req_header_set_single_value(r, name,
-                                                         #name, nil, 0)
+            if not override then
+                error("invalid header value", 3)
```
Comment (thibaultcha) [inline lib/resty/core/request.lua]: I think we should be more descriptive here. We usually are in our other APIs... How about: `"bad value arg: string or table expected, got nil"`. We should also update the `if not value` condition above to better handle truthy/falsy values. So maybe change it to `if value == nil`? The condition in `ngx_resp.add_header()` does this already.

File: lib/resty/core/request.lua [hunk 2]
```diff
@@ -310,16 +314,53 @@ do
...
 
         else
-            if type(value) ~= "string" then
-                value = tostring(value)
+            local sval, sval_len, mvals, mvals_len, buf
+
+            if type(value) == "table" then
+                mvals_len = #value
+                if mvals_len == 0 and not override then
+                    error("invalid header value", 3)
```
Comment (thibaultcha) [inline lib/resty/core/request.lua]: Just like the `ngx_resp.add_header()` API, we should early-exit here, but not consider the argument invalid.

File: lib/resty/core/request.lua [hunk 3]
```diff
@@ -332,6 +373,19 @@ do
...
+    _M.set_req_header = set_req_header
+
+    local orig_func = ngx.req.set_header
+
+    function ngx.req.set_header(name, value)
+        if type(value) == "table" then
+            return orig_func(name, value)
+        end
+
+        local override = true
```
Comment (thibaultcha) [inline lib/resty/core/request.lua]: Better to avoid using a variable here and directly pass the argument. A comment can explain what it is if desired.

File: t/ngx-req.t [hunk 4]
```diff
@@ -0,0 +1,175 @@
+use lib 'lib';
+use Test::Nginx::Socket::Lua;
+use Cwd qw(cwd);
```
Comment (thibaultcha) [inline t/ngx-req.t]: Please use the `t::TestCore` module, just like the `ngx-resp.t` suite.

File: t/ngx-req.t [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+    if (!defined $block->error_log) {
+        $block->set_value("no_error_log", "[error]");
+    }
+
+    if (!defined $block->request) {
+        $block->set_value("request", "GET /t");
+    }
+
+});
```
Comment (thibaultcha) [inline t/ngx-req.t]: Using the `t::TestCore` module will allow to get rid of most of the code duplication here.

File: t/ngx-req.t [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+run_tests();
+
+__DATA__
+
+=== TEST 1: ngx.req.add_header (jitted)
+--- config
+    location = /t {
+        content_by_lua_block {
+            local ngx_req = require "ngx.req"
+            for i = 1, 500 do
```
Comment (thibaultcha) [inline t/ngx-req.t]: Please use the `TEST_NGINX_HOTLOOP` variable when trying to trigger a JIT compilation in tests. E.g. as in `re-find.t`: `for i = 1, $TEST_NGINX_HOTLOOP * 10 do`.

File: t/ngx-req.t [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+            check_invalid_header_value()
+            check_invalid_header_value(nil)
+            check_invalid_header_value({})
+        }
+    }
+--- response_body
+invalid header value
+invalid header value
+invalid header value
```
Comment (thibaultcha) [inline t/ngx-req.t]: Let's also add test cases for boolean values, so we cover our ground with regards to the `if value == nil` condition.

**Comment by thibaultcha** [comment]
Merged!
